src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -381,80 +390,96 @@ entails SolverOptions{..} constraint context hints =
         subclassDictionaryValue dict className index =
           App (Accessor (mkString (superclassName className index)) dict) valUndefined
 
-    solveCoercible :: Environment -> [SourceType] -> Maybe [TypeClassDict]
-    solveCoercible env [a, b] = do
+    solveCoercible :: Environment -> [SourceType] -> [SourceType] -> m (Maybe [TypeClassDict])
+    solveCoercible env kinds [a, b] = runMaybeT $ do
       let tySynMap = typeSynonyms env
           kindMap = types env
-          replaceTySyns = either (const Nothing) Just . replaceAllTypeSynonymsM tySynMap kindMap
+          replaceTySyns = lift . replaceAllTypeSynonymsM tySynMap kindMap
       a' <- replaceTySyns a
       b' <- replaceTySyns b
       -- Solving terminates when the two arguments are the same. Since we
       -- currently don't support higher-rank arguments in instance heads, term
       -- equality is a sufficient notion of "the same".
       if a' == b'
-        then pure [TypeClassDictionaryInScope [] 0 EmptyClassInstance [] C.Coercible [] [] [a, b] Nothing]
+        then pure [TypeClassDictionaryInScope [] 0 EmptyClassInstance [] C.Coercible [] kinds [a, b] Nothing]
         else do
+          lift $ do
+            (_, kind) <- kindOf a'
+            (_, kind') <- kindOf b'
+            unifyKinds kind kind'
```

Comment:
Was this specifically needed? Sorry, I did not mean to imply we needed to unify kinds in `Coercible`, just that kind unification will happen and we need to propagate kind arguments to sub-goals.

Reply:
This yields, I think, better error messages when failing to solve `Coercible` constraints on arguments with different kinds. Compare the error you get when compiling tests/purs/failing/CoercibleKindMismatch.purs with this check:

```
Error found:
in module Main
at ../purs/failing/CoercibleKindMismatch.purs:17:39 - 17:45 (line 17, column 39 - line 17, column 45)

  Could not match kind
        
    Type
        
  with kind
                
    Type -> Type
                

while solving type class constraint
                              
  Prim.Coerce.Coercible Unary 
                        Binary
                              
while checking that type forall (a :: Type) (b :: Type). Coercible a b => a -> b
  is at least as general as type Proxy @(Type -> Type) Unary -> Proxy @(Type -> ...) Binary
while checking that expression coerce
  has type Proxy @(Type -> Type) Unary -> Proxy @(Type -> ...) Binary
in value declaration unaryToBinary

See https://github.com/purescript/documentation/blob/master/errors/KindsDoNotUnify.md for more information,
or to contribute content related to this error.
```

and without:

```
Error found:
in module Main
at ../purs/failing/CoercibleKindMismatch.purs:18:17 - 18:23 (line 18, column 17 - line 18, column 23)

  No type class instance was found for
                                        
    Prim.Coerce.Coercible (Unary t0)    
                          (Binary t0 t1)
                                        

while checking that type forall (a :: Type) (b :: Type). Coercible a b => a -> b
  is at least as general as type Proxy @(Type -> Type) Unary -> Proxy @(Type -> ...) Binary
while checking that expression coerce
  has type Proxy @(Type -> Type) Unary -> Proxy @(Type -> ...) Binary
in value declaration unaryToBinary

See https://github.com/purescript/documentation/blob/master/errors/NoInstanceFound.md for more information,
or to contribute content related to this error.
```

Reply:
In that case I think you should use `elaborateKind` rather than `kindOf`. `elaborateKind` is like a shortcut kind inference for these cases where something is already kind-correct, and all you want to know is the kind rather than a rewritten term.

Reply:
The name is a bit of a misnomer. It should probably be `elaboratedKind` (it returns the kind of an already elaborated term) or `fastKindOf` 😆.

Reply:
Oh that’s why `unifySubstKind` calls `elaborateKind` instead of `kindOf`! Thank you!

Do you think I should also apply the current substitution to the elaborated kinds? I can’t find a situation where this makes any difference.

Reply:
There’s an error with polymorphic kinds now.

This minimal example extracted from tests/purs/passing/Coercible.purs fails to compile:

```purs
module Example where

import Safe.Coerce (coerce)

newtype Phantom1 a b = Phantom1 a

newtype Roles1 a b c = Roles1 (Phantom1 b c)

example :: forall r s t. Roles1 r s t -> s
example = coerce
```

```
Error found:
in module Example
at Example.purs:10:11 - 10:17 (line 10, column 11 - line 10, column 17)

  Internal compiler error:

    Cannot apply type to type: Phantom1 s1

  CallStack (from HasCallStack):
  internalCompilerError, called at src/Language/PureScript/TypeChecker/Kinds.hs:273:3 in purescript-0.13.8-2iUzbEzzqK93DOWR2YKUzM:Language.PureScript.TypeChecker.Kinds
  cannotApplyTypeToType, called at src/Language/PureScript/TypeChecker/Kinds.hs:508:9 in purescript-0.13.8-2iUzbEzzqK93DOWR2YKUzM:Language.PureScript.TypeChecker.Kinds
  elaborateKind, called at src/Language/PureScript/TypeChecker/Kinds.hs:496:11 in purescript-0.13.8-2iUzbEzzqK93DOWR2YKUzM:Language.PureScript.TypeChecker.Kinds
  elaborateKind, called at src/Language/PureScript/TypeChecker/Entailment.hs:407:21 in purescript-0.13.8-2iUzbEzzqK93DOWR2YKUzM:Language.PureScript.TypeChecker.Entailment
  Please report this at https://github.com/purescript/purescript/issues

while solving type class constraint
                                        
  Prim.Coerce.Coercible (Phantom1 s3 t4)
                        s3              
                                        
while checking that type forall (a :: Type) (b :: Type). Coercible a b => a -> b
  is at least as general as type Roles1 @t0 @t1 r2 s3 t4 -> s3
while checking that expression coerce
  has type Roles1 @t0 @t1 r2 s3 t4 -> s3
in value declaration example

where r2 is a rigid type variable
        bound at (line 10, column 11 - line 10, column 17)
      s3 is a rigid type variable
        bound at (line 10, column 11 - line 10, column 17)
      t4 is a rigid type variable
        bound at (line 10, column 11 - line 10, column 17)
      t0 is an unknown type
      t1 is an unknown type

See https://github.com/purescript/documentation/blob/master/errors/InternalCompilerError.md for more information,
or to contribute content related to this error.
```

But adding a kind declaration to `Phantom1` is enough to resolve the error:

```diff
+newtype Phantom1 :: Type -> Type -> Type
 newtype Phantom1 a b = Phantom1 a
```

Reply:
That suggests to me that something is ill-kinded. For example, if you use elaborateKind with `Proxy Id` where there is no kind-application (`Proxy @(Type -> Type) Id`), then it will probably generate an error like that. When it looks up the kind of `Proxy`, it sees it has a `forall`, and not an arrow.

Reply:
It seems that the polymorphic kind of row tails isn’t instantiated under newtypes:

```purs
module Example where

import Safe.Coerce (coerce)

newtype N f = N (f {})

example :: forall f. N f -> f {}
example = coerce
```

```
Error found:
in module Example
at Example.purs:8:11 - 8:17 (line 8, column 11 - line 8, column 17)

  No type class instance was found for
                                                  
    Prim.Coerce.Coercible (f0 (Record ()))        
                          (f0 (Record (() @Type)))
                                                  

while checking that type forall (a :: Type) (b :: Type). Coercible @Type a b => a -> b
  is at least as general as type N f0 -> f0 (Record (() @Type))
while checking that expression coerce
  has type N f0 -> f0 (Record (() @Type))
in value declaration example

where f0 is a rigid type variable
        bound at (line 8, column 11 - line 8, column 17)

See https://github.com/purescript/documentation/blob/master/errors/NoInstanceFound.md for more information,
or to contribute content related to this error.
```

The previous example compiles fine when comparing the types rewritten by `kindOf` in `solveCoercible`.

Is there any other way to wrap row tails in kind applications when unwrapping newtypes?

Reply:
It sounds like this is an error somewhere in the kind-checker then? The types in the environment should always be fully elaborated with their kind-applications. It should never be necessary to do full kind inference during constraint solving.